# Imports

In [44]:
import pandas as pd

The project CSV files are stored in the data/ folder.


In [ ]:
campaña = pd.read_csv('../data/campaña.csv')
clientes = pd.read_csv('../data/clientes.csv')
devoluciones = pd.read_csv('../data/devoluciones.csv')
lineas_pedido = pd.read_csv('../data/lineas_pedido.csv')
pedidos = pd.read_csv('../data/pedidos.csv')
productos = pd.read_csv('../data/productos.csv')
visitas_web = pd.read_csv('../data/visitas_web.csv')


# Initial exploration of the databases

In [50]:
print('='*100)
print('Base de datos: Campaña')
print("="*100)
print(f'Shape : {campaña.shape}')   # (filas, columnas)
display(campaña.head())             # primeras 5 filas
print(campaña.dtypes)               # que tipo tiene cada columna
print(campaña.isnull().sum())       # Cuantos nulos hay por columnas

Base de datos: Campaña
Shape : (4, 8)


,campana_id,nombre,tipo,fecha_inicio,fecha_fin,canal,presupuesto,ventas_ano_anterior
0,1,Rebajas Verano 2024,rebajas,2024-07-01,2024-07-21,instagram,48000.0,512340.0
1,2,Rebajas Invierno 2024,rebajas,2024-01-07,2024-01-28,email,52000.0,498120.0
2,3,Black Friday 2023,black_friday,2023-11-24,2023-11-27,google,61.000 €,405780.0
3,4,Vuelta al Cole 2023,promo,2023-09-01,2023-09-15,instagram,22000.0,176900.0


campana_id               int64
nombre                     str
tipo                       str
fecha_inicio               str
fecha_fin                  str
canal                      str
presupuesto                str
ventas_ano_anterior    float64
dtype: object
campana_id             0
nombre                 0
tipo                   0
fecha_inicio           0
fecha_fin              0
canal                  0
presupuesto            0
ventas_ano_anterior    0
dtype: int64


## Campaign table
- Shape : (4, 8)
- campaign_id : unique identifier for the campaign table. It is the key: it does not repeat (except for export duplicates that must be removed).
- start_date, end_date: Campaign start and end date. Note: column format (str), must be converted to date format. 
- budget: Capital allocated to the campaign. Note: some records contain currency symbols; convert to a number.

In [51]:
print('='*100)
print('Base de datos: Clientes')
print("="*100)
print(f'Shape : {clientes.shape}')
display(clientes.head())
print(clientes.dtypes)
print(clientes.isnull().sum())

Base de datos: Clientes
Shape : (5010, 5)


,cliente_id,fecha_registro,ciudad,genero,edad
0,2371,2019-10-18,bilbao,F,29
1,3537,2021-10-14,Bilbao,M,37
2,2088,2023-10-05,madrid,m,36
3,2769,2018-01-01,Lisboa,F,36
4,1567,2021-06-25,Madrid,M,42


cliente_id        int64
fecha_registro      str
ciudad              str
genero              str
edad              int64
dtype: object
cliente_id          0
fecha_registro    150
ciudad              0
genero              0
edad                0
dtype: int64


## Customers table
- Shape : (5010, 5)
- customer_id: customer identifier. Integer. It is the key: each customer has a unique id. 
- registration_date: Customer registration date. Note: column format (str), must be converted to date format. There are 150 null records. 
- city: City where the registration was made or where the customer lives. Note: text is dirty: "bilbabo", "Bilbao". It must be normalized. 
- gender: customer gender. Note: there are variations. Standardize the column so there are no categories other than F and M.  

In [52]:
print('='*100)
print('Base de datos: devoluciones')
print("="*100)
print(f'Shape : {devoluciones.shape}')
display(devoluciones.head())
print(devoluciones.dtypes)
print(devoluciones.isnull().sum())

Base de datos: devoluciones
Shape : (5558, 5)


,devolucion_id,linea_id,fecha_devolucion,talla,motivo
0,702838,910363,15/07/2024,36,talla incorrecta
1,700079,900324,18/07/2024,XL,no me gusta
2,704821,917455,2024-07-16,m,talla incorrecta
3,701713,906266,17/07/2024,s,no me gusta
4,701621,905938,2024-07-05,36,talla incorrecta


devolucion_id       int64
linea_id            int64
fecha_devolucion      str
talla                 str
motivo                str
dtype: object
devolucion_id       0
linea_id            0
fecha_devolucion    0
talla               0
motivo              0
dtype: int64


## Returns table
- Shape : (5558, 5)
- return_id: unique return identifier. 
- line_id: unique line identifier. Integer. It is the key: it does not repeat. 
- return_date: return date. Note: column format (str), must be converted to date format.
- size: purchased size. Text, and it comes dirty: "M", "m", "Talla M" are the same. It must be normalized.
- reason: return reason. 


In [48]:
print('='*100)
print('Base de datos: lineas pedido')
print("="*100)
print(f'Shape : {lineas_pedido.shape}')
display(lineas_pedido.head())
print(lineas_pedido.dtypes)
print(lineas_pedido.isnull().sum())

Base de datos: lineas pedido
Shape : (20140, 7)


,linea_id,pedido_id,producto_id,talla,cantidad,precio_unitario,descuento_pct
0,900001,500001,1137,39,1,"63,83 €",0
1,900002,500002,1232,XL,1,"26,48 €",0.3
2,900003,500002,1028,L,2,"48,82 €",30%
3,900004,500003,1229,S,1,"38,72 €",0.0
4,900005,500003,1284,S,2,"20,70 €",0%


linea_id           int64
pedido_id          int64
producto_id        int64
talla                str
cantidad           int64
precio_unitario      str
descuento_pct        str
dtype: object
linea_id           0
pedido_id          0
producto_id        0
talla              0
cantidad           0
precio_unitario    0
descuento_pct      0
dtype: int64


## Order lines table
- Shape : (20140, 7)
- line_id : unique line identifier
- order_id : which order the line belongs to. Integer. Points to the orders table.
- product_id : which product it is. Integer. Points to the products table.
- size: purchased size. Text, and it comes dirty: "M", "m", "Talla M" are the same. It must be normalized.
- unit_price: catalog price per unit. It comes as text with "€" and decimal comma ("29,99 €"): convert to number.
- discount_pct : applied discount. It comes in three formats ("20%", 0.2, 20): normalize to a 0-1 fraction.
- Quantity: Quantity purchased. Note: 6 negative values. 

In [47]:
print('='*100)
print('Base de datos: pedidos')
print("="*100)
print(f'Shape : {pedidos.shape}')
display(pedidos.head())
print(pedidos.dtypes)
print(pedidos.isnull().sum())

Base de datos: pedidos
Shape : (8000, 7)


,pedido_id,cliente_id,fecha_pedido,campana_id,canal,ciudad,metodo_pago
0,500001,2290,24/06/2024,NaN,directo,Oporto,tarjeta
1,500002,4937,2024-07-02,1.0,instagram,lisboa,tarjeta
2,500003,544,2024-06-27,NaN,email,Oporto,bizum
3,500004,34,2024-07-12,1.0,email,Zaragoza,paypal
4,500005,4507,2024-07-05,1.0,instagram,Madrid,paypal


pedido_id         int64
cliente_id        int64
fecha_pedido        str
campana_id      float64
canal               str
ciudad              str
metodo_pago         str
dtype: object
pedido_id          0
cliente_id         0
fecha_pedido       0
campana_id      1179
canal              0
ciudad             0
metodo_pago        0
dtype: int64


## Orders table
- Shape : (8000, 7)
- campaign_id : unique identifier for the campaign table. Note: there are 1179 null values identified. 
- order_date: date the order was placed. Note: incorrect format, must be converted to date format. 
- city: City where the registration was made or where the customer lives. Note: text is dirty: "bilbabo", "Bilbao". It must be normalized. 

In [46]:
print('='*100)
print('Base de datos: productos')
print("="*100)
print(f'Shape : {productos.shape}')
display(productos.head())
print(productos.dtypes)
print(productos.isnull().sum())

Base de datos: productos
Shape : (400, 7)


,producto_id,nombre,categoria,temporada,coleccion,coste_unitario,precio_base
0,1001,Vestido Aura,Vestidos,Verano24,Cápsula Sol,12.47,32.57
1,1002,Vestido Iris,VESTIDOS,verano,Atelier,11.01,27.23
2,1003,Vestido Nácar,vestido,verano 2024,Mediterráneo,22.75,55.06
3,1004,Vestido Junio,vestido,Verano24,Básicos,21.83,45.55
4,1005,Vestido Bruma,vestido,verano,Denim,12.20,31.32


producto_id         int64
nombre                str
categoria             str
temporada             str
coleccion             str
coste_unitario    float64
precio_base       float64
dtype: object
producto_id        0
nombre             0
categoria          0
temporada          0
coleccion          0
coste_unitario    12
precio_base        0
dtype: int64


## Products table
- Shape : (400, 7)
- category: Product category. Note: dirty data ' Vestidos',     'VESTIDOS',      'vestido',     'vestidos' must be normalized. 
- season: Season when the product was sold. Note: dirty data 'Verano24', 'verano', 'verano 2024', 'SS24', 'Verano 2024' must be normalized. 
- unit_cost: Unit cost of the product. Note: 12 null values found in the unit_cost column

In [45]:
print('='*100)
print('Base de datos: visitas web')
print("="*100)
print(f'Shape : {visitas_web.shape}')
display(visitas_web.head())
print(visitas_web.dtypes)
print(visitas_web.isnull().sum())

Base de datos: visitas web
Shape : (112, 3)


,fecha,canal,sesiones
0,2024-06-24,email,NaN
1,2024-06-24,instagram,1421.0
2,2024-06-24,google,1078.0
3,2024-06-24,directo,471.0
4,2024-06-25,email,307.0


fecha           str
canal           str
sesiones    float64
dtype: object
fecha       0
canal       0
sesiones    4
dtype: int64


## web_visits table
- Shape : (112, 3)
- date: web visit date. Note: incorrect format (str), convert to date format 
- channel: medium through which the customer visits us. 
- sessions: total number of interactions or continuous visits users make on your website within a given period. Note: there are 4 null values.


# General Analysis

## Campaign table
- Shape : (4, 8)
- campaign_id : unique identifier for the campaign table. It is the key: it does not repeat (except for export duplicates that must be removed).
- start_date, end_date: Campaign start and end date. Note: column format (str), must be converted to date format. 
- budget: Capital allocated to the campaign. Note: some records contain currency symbols; convert to a number.

## Customers table
- Shape : (5010, 5)
- customer_id: customer identifier. Integer. It is the key: each customer has a unique id. 
- registration_date: Customer registration date. Note: column format (str), must be converted to date format. There are 150 null records. 
- city: City where the registration was made or where the customer lives. Note: text is dirty: "bilbabo", "Bilbao". It must be normalized. 
- gender: customer gender. Note: there are variations. Standardize the column so there are no categories other than F and M.  

## Returns table
- Shape : (5558, 5)
- return_id: unique return identifier. 
- line_id: unique line identifier. Integer. It is the key: it does not repeat. 
- return_date: return date. Note: column format (str), must be converted to date format.
- size: purchased size. Text, and it comes dirty: "M", "m", "Talla M" are the same. It must be normalized.
- reason: return reason. Note: there are two categories that say the same thing: "talla incorrecta" and "Talla". 


## Order lines table
- Shape : (20140, 7)
- line_id : unique line identifier
- order_id : which order the line belongs to. Integer. Points to the orders table.
- product_id : which product it is. Integer. Points to the products table.
- size: purchased size. Text, and it comes dirty: "M", "m", "Talla M" are the same. It must be normalized.
- unit_price: catalog price per unit. It comes as text with "€" and decimal comma ("29,99 €"): convert to number.
- discount_pct : applied discount. It comes in three formats ("20%", 0.2, 20): normalize to a 0-1 fraction.

## Orders table
- Shape : (8000, 7)
- campaign_id : unique identifier for the campaign table. Note: there are 1179 null values identified. 
- order_date: date the order was placed. Note: incorrect format, must be converted to date format. 
- city: City where the registration was made or where the customer lives. Note: text is dirty: "bilbabo", "Bilbao". It must be normalized. 

## Products table
- Shape : (400, 7)
- category: Product category. Note: dirty data ' Vestidos',     'VESTIDOS',      'vestido',     'vestidos' must be normalized. 
- season: Season when the product was sold. Note: dirty data 'Verano24', 'verano', 'verano 2024', 'SS24', 'Verano 2024' must be normalized. 
- unit_cost: Unit cost of the product. Note: 12 null values found in the unit_cost column

## web_visits table
- Shape : (112, 3)
- date: web visit date. Note: incorrect format (str), convert to date format 
- channel: medium through which the customer visits us. 
- sessions: total number of interactions or continuous visits users make on your website within a given period. Note: there are 4 null values.
